internal/pkg/util/machine/machine.go

```diff
@@ -0,0 +1,308 @@
+	},
+	{
+		Arch:       "ppc64",
+		Sif:        sif.HdrArchPPC64,
+		Machine:    elf.EM_PPC64,
+		Class:      elf.ELFCLASS32,
+		Endianness: binary.BigEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x02, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x15},
+	},
+	{
+		Arch:       "ppc64le",
+		Sif:        sif.HdrArchPPC64le,
+		Machine:    elf.EM_PPC64,
+		Class:      elf.ELFCLASS64,
+		Endianness: binary.LittleEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x02, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x15, 0x00},
+	},
+	{
+		Arch:       "mips",
+		Sif:        sif.HdrArchMIPS,
+		Compatible: "mips64",
+		Machine:    elf.EM_MIPS,
+		Class:      elf.ELFCLASS32,
+		Endianness: binary.BigEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x01, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x08},
+	},
+	{
+		Arch:       "mipsle",
+		Sif:        sif.HdrArchMIPSle,
+		Compatible: "mips64le",
+		Machine:    elf.EM_MIPS,
+		Class:      elf.ELFCLASS32,
+		Endianness: binary.LittleEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x01, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x08, 0x00},
+	},
+	{
+		Arch:       "mips64",
+		Sif:        sif.HdrArchMIPS64,
+		Machine:    elf.EM_MIPS,
+		Class:      elf.ELFCLASS64,
+		Endianness: binary.BigEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x02, 0x02, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x08},
+	},
+	{
+		Arch:       "mips64le",
+		Sif:        sif.HdrArchMIPS64le,
+		Machine:    elf.EM_MIPS,
+		Class:      elf.ELFCLASS64,
+		Endianness: binary.LittleEndian,
+		ElfMagic:   []byte{0x7F, 0x45, 0x4C, 0x46, 0x02, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x08, 0x00},
+	},
+}
+
+// ArchFromElf returns the architecture string after inspection of the
+// provided elf binary.
+func ArchFromElf(binary string) (string, error) {
+	e, err := elf.Open(binary)
+	if err != nil {
+		return "", fmt.Errorf("failed to open elf binary %s: %s", binary, err)
+	}
+	defer e.Close()
+
+	for _, f := range formats {
+		if f.Machine == e.Machine && f.Class == e.Class && f.Endianness == e.ByteOrder {
+			return f.Arch, nil
+		}
+	}
+
+	return "", ErrUnknownArch
+}
+
+// ArchFromContainer walks through a container filesystem until it
+// find an elf binary to read target architecture from and returns it.
+// If there is no suitable elf binary or if the architecture is not
+// recognized it will return an empty string.
+func ArchFromContainer(container string) string {
+	// fast path if we can get architecture from shell binary
+	shell := fs.EvalRelative("/bin/sh", container)
+	arch, err := ArchFromElf(filepath.Join(shell, container))
+	if err == nil {
+		return arch
+	}
+
```

Comment:
Would it make sense to add a warning/info/verbose here that says that we are not looking at bin/sh? Just to add a little more verbose if a user has multiple arch binaries in a container. Like:

```suggestion
	sylog.Verbosef("Cant find /bin/sh in container, looking at executable files to find architecture...")
```

Or something like that...

Reply:
it doesn't hurt :) :+1: